Product rating: improve preview style when the product doesn't have rating

gigitux opened the pull request:
This PR addresses the feedback raised [here](https://github.com/woocommerce/woocommerce-blocks/pull/9556#issuecomment-1563339572). 

I'm not sure that I followed the BEM convention correctly. Please, double-check it!

My concern is that the preview on the editor side is different from the frontend result. What do you think, @shaunandrews?

|Editor | Frontend |
|--------|--------|
|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/1a6add91-3eff-43b4-a825-66ab389f023b)|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/825ef58d-5239-4326-8280-700f5f2aad53)| 


Fixes #9666 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->


### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/269b46d7-3370-4fb0-9c54-e7f59801e6a9)|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/808f9f4f-39a1-4888-8650-498152e67ed4)|

### Testing

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Edit the Single Product Template (migrate to the blockified template if it isn't).
2. Add the Product Rating block.
3. Ensure that it looks like the image above.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

gigitux committed:
product rating: improve preview style when no rating is visible

roykho commented:
On my end, I don't see the icon next to the `No reviews` text. ![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/1cb95042-661a-4c69-8fb3-2620d2d6fa59)

gigitux commented:
I see it. Could you try to start with a clean build?

roykho commented on assets/js/atomic/blocks/product-elements/rating/style.scss:
```diff
@@ -56,6 +56,37 @@
+			left: 0;
+			right: 0;
+			position: absolute;
+			color: transparent;
```
So the star wasn't showing for me is because of this line. I think you probably want to use opacity?

gigitux replied:
Which browser are you using? I guess that this issue is [this line](https://github.com/woocommerce/woocommerce-blocks/pull/9684/files/3116d79268704f4b7b7f7a2318bb1175536e6513#diff-0ec89a253e28931bdfdbb528c1b9f187e66dad78d2b325059234e94075294732R78).

roykho replied:
Chrome.

gigitux replied:
I tested with Firefox and Chrome and I'm not able to replicate it:

| Firefox | Chrome |
|--------|--------|
|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/66ebbc0f-689e-4e85-96ae-12e888bebbd2)|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/97563901-44d8-4c9f-b11d-3f29700ec26d)|

roykho commented on assets/js/atomic/blocks/product-elements/rating/block.tsx:
```diff
@@ -40,12 +40,35 @@ const getRatingCount = ( product: ProductResponseItem ) => {
+				`${ parentClassName }__product-rating__norating-container`
+			) }
+		>
+			<div
```
Do we need to have divs within divs to achieve this? I feel like we can just have one div and have the two spans next to each other?

gigitux replied:
I preferred this approach to put easily on the same line height.

roykho replied:
You should still be able to achieve same line height with two spans within a div. Did you try using `flex` on the parent container?

gigitux replied:
`wc-block-components-product-rating__norating` div, we applied the custom font to render the start. Adding `flex` to the parent container could influence the other cases. Keep in mind that this is just a temporary solution.

roykho reviewed (changes requested):
Left comment as to why I couldn't see the star. Also some suggestions.

kmanijak commented on assets/js/atomic/blocks/product-elements/rating/block.tsx:
```diff
@@ -40,12 +40,35 @@ const getRatingCount = ( product: ProductResponseItem ) => {
+		<div
+			className={ classnames(
+				'wc-block-components-product-rating__norating-container',
+				`${ parentClassName }__product-rating__norating-container`
```
> I'm not sure that I followed the BEM convention correctly. Please, double-check it!

BEM convention allows only one element nesting, so according to it, this structure: `block__element__element` is not recommended (e.g. https://scalablecss.com/bem-nesting-grandchild-elements/).

I don't know the exact purpose of the class in here, but to my best knowledge it should be:

```suggestion
				`${ parentClassName }-product-rating__norating-container`
```

gigitux replied:
Thanks for the suggestion!

kmanijak reviewed (approved):
This works as expected!

> My concern is that the preview on the editor side is different from the frontend result.

I feel similar and TBH I think we could show the same content (⭐ No Reviews) in the frontend - it looks nice and makes the layout consistent, so the rating section keeps the same height.

I found one bug, but not related to this PR: all the rating in the reviews section shows the average rating (not reproducible in WC Core).

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9666-product-rating-block-show-no-reviews-label-rather-than-empty-starts-when-product-has-no-reviews

gigitux committed:
fix class naming

gigitux commented:
> This works as expected!
> 
> >

@roykho, could you check again if you can't see it? My suggestion is to try in incognito mode. Maybe you have some extensions that create a conflict?

roykho commented:
Ok I found out that is is theme related. If you try `Tsubaki` theme, you will see that the icon doesn't show. So we might need to account for this as other themes may produce similar issues. When I switched to TTT theme, I see it however it is partially cut off on the right side. See below (your screen shot also has it cut off):
![xiV82r.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/b7bd3d32-1446-4894-8073-a5a64c072e47)

gigitux committed:
fix cutted star

gigitux committed:
change variable

gigitux commented:
> Ok I found out that is is theme related. If you try `Tsubaki` theme, you will see that the icon doesn't show. So we might need to account for this as other themes may produce similar issues.

Oh, nice! I changed the variable. In this way, the border of the star will use the same color of the text 58350ae25d42929a8c027478269cfb64809626c4

> When I switched to TTT theme, I see it however it is partially cut off on the right side. See below (your screen shot also has it cut off)

Nice catch! Fixed with d622b94c931ad479ba4b3b366f11ae800d61ed80

roykho commented:
With the latest changes, now the icon is not showing in Twenty Twenty Three theme (shows in Tsubaki). Also I think I missed it earlier but the left side of the icon is cut off as well:
![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/e636d3c7-d992-470f-af88-6999b1467ebb)

gigitux committed:
fix cut border

gigitux commented:
> With the latest changes, now the icon is not showing in Twenty Twenty Three theme (shows in Tsubaki). Also I think I missed it earlier but the left side of the icon is cut off as well: ![file.png](https://user-images.githubusercontent.com/2132595/243393499-e636d3c7-d992-470f-af88-6999b1467ebb.png)

Sorry, a lot for this back and forth 😞. Block themes use different CSS variables. Since this is just a temporary solution and will not be visible on the frontend I used the `--wp--preset--color--black` variable with a fallback to #000 if it is not available.

Also, I centered the star, so it should not be cut anymore.

roykho reviewed (approved):
Thanks for addressing all the feedbacks. I tested in both TTT and Tsubaki theme and now the star appears in both themes and also it is no longer cut off. Good job!
